includes/modules/management/add/class-wcs-att-manage-add-product.php

```diff
@@ -0,0 +1,437 @@
+
+		remove_action( 'woocommerce_my_subscriptions_actions', array( __CLASS__, 'button_template' ) );
+	}
+
+	/**
+	 * 'Add to subscription' view -- 'Add' button template.
+	 *
+	 * @param  WC_Subscription  $subscription
+	 */
+	public static function button_template( $subscription ) {
+
+		wc_get_template( 'single-product/product-add-to-subscription-button.php', array(
+			'subscription_id' => $subscription->get_id()
+		), false, WCS_ATT()->plugin_path() . '/templates/' );
+	}
+
+	/*
+	|--------------------------------------------------------------------------
+	| Ajax Handlers
+	|--------------------------------------------------------------------------
+	*/
+
+	/**
+	 * Load all user subscriptions matching a product + scheme key (known billing period and interval).
+	 *
+	 * @return void
+	 */
+	public static function load_matching_subscriptions() {
+
+		$failure = array(
+			'result' => 'failure',
+			'html'   => ''
+		);
+
+		// User must be logged in.
+		if ( ! is_user_logged_in() ) {
+			wp_send_json( $failure );
+		}
+
+		$product_id = ! empty( $_POST[ 'product_id' ] ) ? absint( $_POST[ 'product_id' ] ) : false;
+		$scheme_key = ! empty( $_POST[ 'subscription_scheme' ] ) ? wc_clean( $_POST[ 'subscription_scheme' ] ) : false;
+
+		$product = wc_get_product( $product_id );
+
+		if ( ! $product ) {
+			wp_send_json( $failure );
+		}
+
+		// Keep the sneaky folks out.
+		if ( ! WCS_ATT_Product::supports_feature( $product, 'subscription_management_add_to_subscription' ) ) {
+			wp_send_json( $failure );
+		}
+
+		$scheme = WCS_ATT_Product_Schemes::get_subscription_scheme( $product, 'object', $scheme_key );
+
+		// We expect a scheme key to be posted when it's only possible to add the product to matching subscriptions.
+		if ( ! $scheme && WCS_ATT_Product::supports_feature( $product, 'subscription_management_add_to_matching_subscription' ) ) {
+			wp_send_json( $failure );
+		}
+
+		/**
+		 * 'wcsatt_subscriptions_matching_product' filter.
+		 *
+		 * Last chance to filter matched subscriptions.
+		 *
+		 * @param  array                $matching_subscriptions
+		 * @param  WC_Product           $product
+		 * @param  WCS_ATT_Scheme|null  $scheme
+		 */
+		$matching_subscriptions = apply_filters( 'wcsatt_subscriptions_matching_product', WCS_ATT_Manage_Add::get_matching_subscriptions( $scheme ), $product, $scheme );
+
+		ob_start();
+
+		/**
+		 * 'wcsatt_add_product_to_subscription_html' action.
+		 *
+		 * @param  array                $matching_subscriptions
+		 * @param  WC_Product           $product
+		 * @param  WCS_ATT_Scheme|null  $scheme
+		 *
+		 * @hooked WCS_ATT_Manage_Add_Product::matching_subscriptions_template - 10
+		 */
+		do_action( 'wcsatt_add_product_to_subscription_html', $matching_subscriptions, $product, $scheme );
+
+		$html = ob_get_clean();
+
+		if ( ! $html ) {
+			$result = $failure;
+		} else {
+			$result = array(
+				'result' => 'success',
+				'html'   => $html
+			);
+		}
+
+		wp_send_json( $result );
```

Comment:
The 10 lines above are identical to the lines at the end of `WCS_ATT_Manage_Add_Cart::load_matching_subscriptions()`. Seems like a shared parent handling the start and end of this method, with the actual logic abstracted into a child method implemented by `WCS_ATT_Manage_Add_Cart` and `WCS_ATT_Manage_Add_Product` would save quite a bit of duplicate code, so is worth the effort.